tests/formparams/form_params_test.go

```diff
@@ -0,0 +1,383 @@
+// rejected. Go will handle a) as a valid request and c) as an invalid requet, as the following two tests demonstrate
+
+func TestFormParametersValidDuplicateContentLength(t *testing.T) {
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" +
+		"Content-Length: 13\r\n" +
+		"Content-Length: 13\r\n" +
+		"\r\n" +
+		"veggie=potato&veggie=a\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		want := []string{"13"}
+		got := req.Header["Content-Length"]
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.Header: got %v, want %v, diff (-want +got): \n%s", got, want, diff)
+		}
+		req.ParseForm()
+		if want := []string{"potato"}; !cmp.Equal(want, req.Form["veggie"]) {
+			t.Errorf(`Form["veggie"]: got %v, want %v`, req.Form["veggie"], want)
+		}
+
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(status200OK)) {
+		t.Errorf("response status: got %s, want %s", resp, status200OK)
+	}
+
+}
+
+func TestFormParametersInvalidDuplicateContentLength(t *testing.T) {
+	tests := []struct {
+		name string
+		req  []byte
+	}{
+		{
+			name: "Two Content-Length Headers",
+			req: []byte("POST / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				"Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n" +
+				"Content-Length: 13\r\n" +
+				"Content-Length: 12\r\n" +
+				"\r\n" +
+				"veggie=potato\r\n" +
+				"\r\n"),
+		},
+		{
+			name: "List of same value in Content-Length Header",
+			req: []byte("POST / HTTP/1.1\r\n" +
+				"Host: localhost:8080\r\n" +
+				"Content-Type: application/x-www-form-urlencoded; charset=utf-8\r\n" +
+				"Content-Length: 13, 13\r\n" +
+				"\r\n" +
+				"veggie=potato\r\n" +
+				"\r\n"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), test.req, func(req *http.Request) {
+				t.Error("expected handler not to be called.")
+			})
+			if err != nil {
+				t.Fatalf("MakeRequest(): got %v, want nil", err)
+			}
+			if !bytes.HasPrefix(resp, []byte(status400BadReq)) {
+				t.Errorf("response status: got %s, want %s", resp, status400BadReq)
+			}
+		})
+	}
+}
+
+// Test whether form parameters with Content-Type:
+// application/x-www-form-urlencoded that break percent-encoding will be
+// ignored and following valid query parameters will be discarded
+func TestFormParametersBreakUrlEncoding(t *testing.T) {
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded\r\n" +
+		"Content-Length: 22\r\n" +
+		"\r\n" +
+		"veggie=%sc&fruit=apple\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if values := req.Form["veggie"]; len(values) != 0 {
```

Comment:
Last place where you should use `cmp`.

Reply:
Resolved in [4504136bee3f68b1e4b142e96c26e322208484e7](url)